File: 908_smallest_range_i/smallest_range_i.cpp
/*
 * @Date: 2022-04-30 08:16:37
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-04-30 08:28:25
 * @FilePath: /algorithm/908_smallest_range_i/smallest_range_i.cpp
 */

#include <algorithm>
#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
  int smallestRangeI(vector<int> nums, int k) {
    int minNum = *min_element(nums.begin(), nums.end());
    int maxNum = *max_element(nums.begin(), nums.end());
    return maxNum - minNum <= 2 * k ? 0 : maxNum - minNum - 2 * k;
  }
};

int main() {
  assert(Solution().smallestRangeI({1}, 0) == 0);
  assert(Solution().smallestRangeI({0, 10}, 2) == 6);
  assert(Solution().smallestRangeI({1, 3, 6}, 3) == 0);
}

File: golang/908_smallest_range_i/smallest_range_i.go
/*
 * @Date: 2022-04-30 08:16:44
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-04-30 08:27:27
 * @FilePath: /algorithm/908_smallest_range_i/smallest_range_i.go
 */

package main

func smallestRangeI(nums []int, k int) int {
	minNum, maxNum := nums[0], nums[0]
	for _, num := range nums[1:] {
		if num < minNum {
			minNum = num
		} else if num > maxNum {
			maxNum = num
		}
	}
	ans := maxNum - minNum - 2*k
	if ans > 0 {
		return ans
	}
	return 0
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	assert(smallestRangeI([]int{1}, 0) == 0)
	assert(smallestRangeI([]int{0, 10}, 2) == 6)
	assert(smallestRangeI([]int{1, 3, 6}, 3) == 0)
}
